src/ffi/mod.rs

```diff
@@ -29,182 +30,93 @@
-pub use self::structseq::*;
-pub use self::sysmodule::*;
-pub use self::traceback::*;
-pub use self::tupleobject::*;
-pub use self::typeslots::*;
-pub use self::unicodeobject::*;
-pub use self::warnings::*;
-pub use self::weakrefobject::*;
-
-mod abstract_;
-// skipped asdl.h
-// skipped ast.h
-mod bltinmodule;
-mod boolobject;
-mod bytearrayobject;
-mod bytesobject;
-// skipped cellobject.h
-mod ceval;
-// skipped classobject.h
-mod code;
-mod codecs;
-mod compile;
-mod complexobject;
-#[cfg(all(Py_3_8, not(Py_LIMITED_API)))]
-mod context; // It's actually 3.7.1, but no cfg for patches.
+mod datetime;
+
 #[cfg(not(Py_LIMITED_API))]
-pub(crate) mod datetime;
-mod descrobject;
-mod dictobject;
-// skipped dynamic_annotations.h
-mod enumobject;
-// skipped errcode.h
-mod eval;
-// skipped exports.h
-mod fileobject;
-mod fileutils;
-mod floatobject;
-// skipped empty frameobject.h
+pub use datetime::*;
+
 #[cfg(not(Py_LIMITED_API))]
-pub(crate) mod funcobject;
-// skipped genericaliasobject.h
+use {libc::c_int, std::ops::Deref};
+
+//  reexport raw bindings exposed in pyo3_ffi
+pub use pyo3_ffi::*;
+
+/// Helper to enable #\[pymethods\] to see the workaround for __ipow__ on Python 3.7
+#[doc(hidden)]
+pub use crate::impl_::pymethods::ipowfunc;
+
+// Shadow raw ffi bindings for backwards compatability
+
 #[cfg(not(Py_LIMITED_API))]
-mod genobject;
-mod import;
-// skipped interpreteridobject.h
-mod intrcheck;
-mod iterobject;
-mod listobject;
-// skipped longintrepr.h
-mod longobject;
-pub(crate) mod marshal;
-mod memoryobject;
-mod methodobject;
-mod modsupport;
-mod moduleobject;
-// skipped namespaceobject.h
-mod object;
-mod objimpl;
-// skipped odictobject.h
-// skipped opcode.h
-// skipped osdefs.h
-mod osmodule;
-// skipped parser_interface.h
-// skipped patchlevel.h
-// skipped picklebufobject.h
-// skipped pyctype.h
-// skipped py_curses.h
+pub static PyDateTimeAPI: _PyDateTimeAPI_impl = _PyDateTimeAPI_impl {};
+
 #[cfg(not(Py_LIMITED_API))]
-mod pyarena;
-mod pycapsule;
-// skipped pydecimal.h
-// skipped pydtrace.h
-mod pyerrors;
-// skipped pyexpat.h
-// skipped pyfpe.h
-mod pyframe;
-mod pyhash;
-mod pylifecycle;
-// skipped pymacconfig.h
-// skipped pymacro.h
-// skipped pymath.h
-mod pymem;
-mod pyport;
-mod pystate;
-mod pythonrun;
-// skipped pystrhex.h
-// skipped pystrcmp.h
-mod pystrtod;
-// skipped pythread.h
-// skipped pytime.h
-mod rangeobject;
-mod setobject;
-mod sliceobject;
-mod structseq;
-mod sysmodule;
-mod traceback;
-// skipped tracemalloc.h
-mod tupleobject;
-mod typeslots;
-mod unicodeobject;
-mod warnings;
-mod weakrefobject;
-
-// Additional headers that are not exported by Python.h
-pub mod structmember;
-
-// "Limited API" definitions matching Python's `include/cpython` directory.
+pub static PyDateTime_TimeZone_UTC: _PyDateTime_TimeZone_UTC_impl = _PyDateTime_TimeZone_UTC_impl {
+    inner: &PyDateTimeAPI,
+};
+
 #[cfg(not(Py_LIMITED_API))]
-mod cpython;
+macro_rules! reexport_with_autoinit {
```

Comment:
These "autoinit" helpers are very useful for the safe wrapper, but we should explore whether they should instead be moved to be private features of `types/datetime.rs`.

Reply:
I have addressed this. There is no more special datetime handeling in pyo3::ffi compared to pyo3::ffi anymore. See above for details